builtin/v11/migration/miner.go

```diff
@@ -0,0 +1,323 @@
+				case amt.Modify:
+					sectorNo := abi.SectorNumber(change.Key)
+					info, found, err := inSectors.Get(sectorNo)
+					if err != nil {
+						return cid.Undef, xerrors.Errorf("failed to get sector %d in inSectors: %w", sectorNo, err)
+					}
+
+					if !found {
+						return cid.Undef, xerrors.Errorf("didn't find sector %d in inSectors", sectorNo)
+					}
+
+					if err := prevOutSectors.Set(change.Key, migrateSectorInfo(*info)); err != nil {
+						return cid.Undef, xerrors.Errorf("failed to set migrated sector %d in prevOutSectors", sectorNo)
+					}
+				}
+			}
+
+			outArray = prevOutSectors.Array
+		} else {
+			// first time we're doing this, do all the work
+			outArray, err = migrateSectorsFromScratch(ctx, store, inArray)
+			if err != nil {
+				return cid.Undef, xerrors.Errorf("failed to migrate sectors from scratch: %w", err)
+			}
+		}
+
+		outRoot, err := outArray.Root()
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("error writing new sectors AMT: %w", err)
+		}
+
+		if err = cache.Write(migration.MinerPrevSectorsInKey(minerAddr), inRoot); err != nil {
+			return cid.Undef, xerrors.Errorf("failed to write inkey to cache: %w", err)
+		}
+
+		if err = cache.Write(migration.MinerPrevSectorsOutKey(minerAddr), outRoot); err != nil {
+			return cid.Undef, xerrors.Errorf("failed to write inkey to cache: %w", err)
+		}
+
+		return outRoot, nil
+	})
+}
+
+func migrateSectorsFromScratch(ctx context.Context, store adt10.Store, inArray *adt10.Array) (*adt11.Array, error) {
+	outArray, err := adt11.MakeEmptyArray(store, miner11.SectorsAmtBitwidth)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to construct new sectors array: %w", err)
+	}
+
+	var sectorInfo miner10.SectorOnChainInfo
+	if err = inArray.ForEach(&sectorInfo, func(k int64) error {
+		return outArray.Set(uint64(k), migrateSectorInfo(sectorInfo))
+	}); err != nil {
+		return nil, err
+	}
+
+	return outArray, err
+}
+
+func (m minerMigrator) migrateDeadlines(ctx context.Context, store adt10.Store, cache migration.MigrationCache, deadlines cid.Cid) (cid.Cid, error) {
+	if deadlines == m.emptyDeadlinesV10 {
+		return m.emptyDeadlinesV11, nil
+	}
+
+	var inDeadlines miner10.Deadlines
+	err := store.Get(store.Context(), deadlines, &inDeadlines)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	var outDeadlines miner11.Deadlines
+	for i, c := range inDeadlines.Due {
+		if c == m.emptyDeadlineV10 {
+			outDeadlines.Due[i] = m.emptyDeadlineV11
+		} else {
+			var inDeadline miner10.Deadline
+			if err = store.Get(ctx, c, &inDeadline); err != nil {
+				return cid.Undef, err
+			}
+
+			outSectorsSnapshotCid, err := cache.Load(migration.SectorsAmtKey(inDeadline.SectorsSnapshot), func() (cid.Cid, error) {
+				inSectorsSnapshot, err := adt10.AsArray(store, inDeadline.SectorsSnapshot, miner10.SectorsAmtBitwidth)
+				if err != nil {
+					return cid.Undef, err
+				}
+
+				outSectorsSnapshot, err := migrateSectorsFromScratch(ctx, store, inSectorsSnapshot)
```

Comment:
Time to bring this back up: I don't understand why we go to the trouble of AMT diffing for sectors tree but not for sector snapshots.  Have we confirmed that not much time is spent here?  I think we should confirm that and if we actually do spend a decent fraction of time here design something nice.

I think we can diff against the latest sector tree assuming the cache for this miner address is populated.  Because this means that snapshot has changed since pre migration which means that the snapshot tree will be very close to the currently cached sectors AMT.

It will likely also make the premigration a lot faster if we do sectors tree first and then diff all of these against it.  Though this depends on AMT diffing performance which maybe I'm overestimating.

I'm suspicious we can do all of this with 0 caching overhead by just referencing already cached values.  

Why do I think its worth it now? If we do this now we can either find there's no big speed up and drop forever or find there is a nice speedup and finish this now and reuse on all subsequent sector info migrations (like we are doing with the nice diffing code on sectors AMT)

Reply:
I tried this out and empirically we performed *worse* on the premigration.  Unclear why unless AMT diffing is generally inefficient.  Its worth continuing to understand this in the future but for now leaving this is the way to go.